dev/ColorPicker/docs/ColorPickerArchitecture.md

```diff
@@ -0,0 +1,72 @@
+
+### Rendering the Spectrum
+
+There are presently two different techniques used for generating/rendering the color spectrum depending on the version of Windows.
+
+ 1. On Windows version 1607 (RS1) and before, the spectrum is rendered as an array of individual bytes, converted to a WritableBitmap, set as the source of an ImageBrush and then finally set as the background/fill of a spectrum rectangle. 
+ 2. On Windows version 1703 (RS2) and later, the spectrum images are placed in a loaded image surface, which is then put into a SpectrumBrush. This technique uses the Composition APIs.
+
+Rendering is done asynchronously for performance. However, it still can take 10s of milliseconds to re-render the spectrum. This time becomes negligible as the spectrum does not need to re-render itself as the selected color changes. A special trick is used as documented in the `ColorSpectrum.cpp` source:
+
+> As the user perceives it, every time the third dimension not represented in the ColorSpectrum changes, the ColorSpectrum will visually change to accommodate that value.  For example, if the ColorSpectrum handles hue and luminosity, and the saturation externally goes from 1.0 to 0.5, then the ColorSpectrum will visually change to look more washed out to represent that third dimension's new value. Internally, however, we don't want to regenerate the ColorSpectrum bitmap every single time this happens, since that's very expensive. In order to make it so that we don't have to, we implement an optimization where, rather than having only one bitmap, we instead have multiple that we blend together using opacity to create the effect that we want. In the case where the third dimension is saturation or luminosity, we only need two: one bitmap at the minimum value of the third dimension, and one bitmap at the maximum.  Then we set the second's opacity at whatever the value of the third dimension is - e.g., a saturation of 0.5 implies an opacity of 50%. In the case where the third dimension is hue, we need six: one bitmap corresponding to red, yellow, green, cyan, blue, and purple. We'll then blend between whichever colors our hue exists between - e.g., an orange color would use red and yellow with an opacity of 50%. This optimization does incur slightly more startup time initially since we have to generate multiple bitmaps at once instead of only one, but the running time savings after that are *huge* when we can just set an opacity instead of generating a brand new bitmap.
+
+## SpectrumBrush
+
+TODO
+
+## Ideas for Future Implementations
```

Comment:
I am not sure if this is better kept in an issue instead of the documentation of the control. This section really doesn't talk about the existing control but rather what things one could potentially change. Especially things like "X should be named Y instead" seems a bit strange in the documentation.

Reply:
Lessons learned always seem to get lost. Adding them here ensures they survive with the code. Im adamant this section stays.

Reply:
Usually there is a specific document or even database for mistakes made in project(s). "Lessons learned" usually lives in a separate area that is easier to search through instead of the architecture documentation. @ranjeshj @MikeHillberg What are your thoughts on this?

Reply:
@chingucoding We keep disagreeing on almost everything which is making our discussions quite challenging. If any lesson's learned exists it is internal to Microsoft and not visible publicly. Then, as code goes open source like this, it is lost. It's best to keep it in code in this case. Of course you are right and traditionally there is a separate lesson's learned database though.

Reply:
I did not imply that we should have it be in an internal or private database, I am just trying to say that I think it should be in a separate document, not next to the actual current documentation of the control. But that's just my opinion ...